vspace/src/main/java/edu/asu/diging/vspace/web/staff/SpaceOverviewController.java

```diff
@@ -0,0 +1,172 @@
+            spaceList.forEach(s -> {
+                ObjectNode node = mapper.createObjectNode();
+                node.put("name", s.getName());
+                node.put("link", request.getContextPath() + "/staff/space/" + s.getId());
+                node.put("img", request.getContextPath() + "/api/image/" + s.getImage().getId());
+                node.put("isModule", false);
+                List<String> spaceAllLink = new ArrayList<>();
+                List<IModuleLinkDisplay> spaceToModulelinksList = AllSpaceToModulelinksListMap.get(s.getId());
+                List<ISpaceLinkDisplay> spaceLinks = null;
+                if (s.isShowUnpublishedLinks()) {
+                    spaceLinks = AllSpaceToSpacelinksListMap.get(s.getId());
+                } else {
+                    spaceLinks = AllSpaceLinkForGivenOrNullSpaceStatustMap.get(s.getId());
+                }
+                List<ISpaceLinkDisplay> spaceToSpaceLinksList = null;
+                if (spaceLinks != null) {
+                    spaceToSpaceLinksList = spaceLinks.stream().filter(spaceLinkDisplayObj -> !spaceLinkDisplayObj
+                            .getLink().getTargetSpace().isHideIncomingLinks()).collect(Collectors.toList());
+                }
+                if (spaceToModulelinksList != null && !spaceToModulelinksList.isEmpty()) {
+
+                    spaceToModulelinksList.forEach(link -> {
+                        spaceAllLink.add(link.getLink().getModule().getId());
+                    });
+                }
+                if (spaceToSpaceLinksList != null && !spaceToSpaceLinksList.isEmpty()) {
+
+                    spaceToSpaceLinksList.forEach(link -> {
+                        spaceAllLink.add(link.getLink().getTarget().getId());
+                    });
+                }
+                spaceLinkMap.put(s.getId(), spaceAllLink);
+                allNodeList.add(s.getId());
+                nodeArray.add(node);
+            });
+        }
+        if (moduleList != null && !moduleList.isEmpty()) {
+            moduleList.forEach(m -> {
+                ObjectNode node = mapper.createObjectNode();
+                node.put("name", m.getName());
+                node.put("link", request.getContextPath() + "/staff/module/" + m.getId());
+                node.put("img", "");
+                node.put("isModule", true);
+                allNodeList.add(m.getId());
+                nodeArray.add(node);
+            });
+        }
+        model.addAttribute("overViewNode", mapper.writeValueAsString(nodeArray));
+        spaceLinkMap.forEach((k, v) -> {
+            if (!v.isEmpty()) {
+                v.forEach(e -> {
+                    ObjectNode node = mapper.createObjectNode();
+                    node.put("target", allNodeList.indexOf(e));
+                    node.put("source", allNodeList.indexOf(k));
+                    linkArray.add(node);
+                });
+            }
+        });
```

Comment:
at least one of these loops should be a proper for loop; this nesting of `forEach` does not help with readability; it has the opposite effect. and give the variables proper names;